modules/3d/test/test_normal.cpp

```diff
@@ -396,12 +369,305 @@ class RgbdPlaneTest
+        // depthTo3d(srcDepth, intr, srcCloud, srcMask);
+        depthTo3d(srcDepth, intr, srcCloud);
+        Scalar qnan = Scalar::all(std::numeric_limits<double>::quiet_NaN());
+        srcCloud.setTo(qnan, ~srcMask);
+        srcDepth.setTo(qnan, ~srcMask);
+
+        // For further result comparison
+        srcNormals.setTo(qnan, ~srcMask);
+
+        // TODO: test cases like above
+        Ptr<RgbdNormals> normalsComputer = RgbdNormals::create(1080, 1920, depth, intr, 5, diffThreshold, method);
+        normalsComputer->cache();
+
+        Mat dstNormals;
+        //TODO: LINEMOD can accept 16U and cloud instead of depth, fix it
+        if (normalsComputer->getMethod() == RgbdNormals::RGBD_NORMALS_METHOD_LINEMOD)
+        {
+            normalsComputer->apply(srcDepth, dstNormals);
+        }
+        else
+            normalsComputer->apply(srcCloud, dstNormals);
+
+        Result res;
+
+        // Remove 4th channel from dstNormals
+        Mat newDstNormals;
+        std::vector<Mat> dstNormalsCh;
+        split(dstNormals, dstNormalsCh);
+        dstNormalsCh.resize(3);
+        merge(dstNormalsCh, newDstNormals);
+        dstNormals = newDstNormals;
+
+        Mat dstMask(dstNormals.size(), CV_8U);
+        for (int y = 0; y < dstMask.rows; y++)
+        {
+            uchar* dstMaskRow = dstMask.ptr<uchar>(y);
+            if (depth == CV_32F)
+            {
+                Vec3f *dstrow = dstNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *dstrow = dstNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+        }
+        // div by 8 because uchar is 8-bit
+        res.maskl2 = cv::norm(dstMask, srcMask, NORM_HAMMING)/8;
+
+        // Flipping Y and Z to correspond to srcNormals
+
+        for (int y = 0; y < dstNormals.rows; y++)
+        {
+            if (depth == CV_32F)
+            {
+                Vec3f *nrmrow = dstNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < dstNormals.cols; x++)
+                {
+                    Vec3f n = nrmrow[x];
+                    n[1] = -n[1];
+                    n[2] = -n[2];
+                    nrmrow[x] = n;
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *nrmrow = dstNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < dstNormals.cols; x++)
+                {
+                    Vec3d n = nrmrow[x];
+                    n[1] = -n[1];
+                    n[2] = -n[2];
+                    nrmrow[x] = n;
+                }
+            }
+        }
```

Comment:
I propose to add axes flip to 3 module itself as function. It's useful commodity not only for the tests.

Reply:
Can this be achieved by some MatExpr-like expression?

Reply:
Theoretically yes, but OpenCV has design issues in MatExpr resolution like this: https://github.com/opencv/opencv/issues/16538. Please pay attention.